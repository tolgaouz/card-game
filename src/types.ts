import { TransitionStatus } from 'react-transition-group';

export enum CardKind {
  CLUBS = 'Clubs',
  DIAMONDS = 'Diamonds',
  HEARTS = 'Hearts',
  SPADES = 'Spades',
}
/*
Denote non-number types with numbers as well,
1 --> Ace
11 --> Jack
12 --> Queen
13 --> King
*/
export type CardNumber = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;
export enum CardKindColorCSS {
  RED = 'text-redCard',
  BLACK = 'text-black',
}
export type Card = { kind: CardKind; number: CardNumber };

export type Ace = { number: 1; kind: CardKind };
export type King = { number: 13; kind: CardKind };
export type Queen = { number: 12; kind: CardKind };
export type Jack = { number: 11; kind: CardKind };

export interface CardGameState {
  deck: Card[];
  hand: Card[];
  acesInDeck: Ace[];
  finished?: boolean;
  userWon?: boolean;
}

export type TransitionStyle = Exclude<TransitionStatus, 'unmounted'>;
