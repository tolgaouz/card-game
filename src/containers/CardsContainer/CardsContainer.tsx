import clsx from 'clsx';
import { CSSProperties } from 'react';
import { useMediaQuery } from 'react-responsive';
import { TransitionGroup, Transition } from 'react-transition-group';
import Card from '../../components/Card';
import MediaQueries from '../../mediaQueries';
import { CardGameState, TransitionStyle } from '../../types';

interface CardsContainerProps extends React.HTMLProps<HTMLDivElement> {
  cards: CardGameState['hand'];
}

const CardsContainer: React.FC<CardsContainerProps> = ({ cards, className }) => {
  // Should be in-sync with the theme object
  const isDesktop = useMediaQuery({
    query: `(min-width: ${MediaQueries.md})`,
  });
  const duration = 300;
  return (
    <TransitionGroup
      className={clsx(
        'flex my-6 md:px-8 md:my-32 flex-wrap gap-x-3 gap-y-5 md:gap-5 justify-center w-full',
        className
      )}
    >
      {cards.map((card, idx) => {
        // Taking the card in the middle as center point,
        // spread the cars along a circular perimeter
        // Tailwind won't work with dynamically constructed class names,
        // for reference: https://tailwindcss.com/docs/content-configuration#dynamic-class-names
        const centerIndex = Math.floor(cards.length / 2);
        const defaultStyle: CSSProperties = {
          transition: `all ${duration}ms ease-in-out`,
          transformOrigin: '50% 0',
        };
        let transitionStyles: {
          [k in TransitionStyle]?: CSSProperties;
        } = {
          entering: { opacity: 0 },
          entered: {
            opacity: 1,
            transform: `translateY(20px)`,
          },
          exiting: { display: 'none' },
          exited: { display: 'none' },
        };
        // Dont need to apply circular transform if cards length is smaller than 3 or
        // mobile view
        if (cards.length > 3 && isDesktop) {
          transitionStyles = {
            entering: { opacity: 0 },
            entered: {
              opacity: 1,
              transform: `rotate(${(centerIndex - idx) * 6}deg) translateY(${
                Math.cos((centerIndex - idx) * 12) * 80
              }px)`,
            },
            exiting: { display: 'none' },
            exited: { display: 'none' },
          };
        }
        return (
          <Transition key={`${card.kind}_${card.number}`} in timeout={duration}>
            {(state) => (
              <Card
                style={{
                  ...defaultStyle,
                  ...transitionStyles[state as TransitionStyle],
                }}
                kind={card.kind}
                number={card.number}
              />
            )}
          </Transition>
        );
      })}
    </TransitionGroup>
  );
};

export default CardsContainer;
