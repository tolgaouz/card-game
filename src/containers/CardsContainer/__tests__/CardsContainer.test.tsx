import { render } from '@testing-library/react';
import { Card, CardKind } from '../../../types';
import CardsContainer from '../CardsContainer';

const cards: Card[] = [
  { kind: CardKind.CLUBS, number: 2 },
  { kind: CardKind.CLUBS, number: 3 },
  { kind: CardKind.CLUBS, number: 4 },
  { kind: CardKind.CLUBS, number: 5 },
  { kind: CardKind.CLUBS, number: 6 },
];

describe('Circular Cards View', () => {
  it('renders correctly', () => {
    const container = render(<CardsContainer cards={cards} />);
    expect(container).toMatchSnapshot();
  });
});
