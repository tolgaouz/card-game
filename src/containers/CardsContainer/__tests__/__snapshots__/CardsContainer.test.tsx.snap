// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circular Cards View renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex my-6 md:px-8 md:my-32 flex-wrap gap-x-3 gap-y-5 md:gap-5 justify-center w-full"
      >
        <div
          class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
          style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
        >
          <div
            class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
            data-testid="card-text"
          >
            2
            <img
              alt="2 of Clubs"
              class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
              src="Clubs.svg"
            />
          </div>
          <img
            alt="2 of Clubs"
            class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
            data-testid="card-symbol"
            src="Clubs.svg"
          />
        </div>
        <div
          class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
          style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
        >
          <div
            class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
            data-testid="card-text"
          >
            3
            <img
              alt="3 of Clubs"
              class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
              src="Clubs.svg"
            />
          </div>
          <img
            alt="3 of Clubs"
            class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
            data-testid="card-symbol"
            src="Clubs.svg"
          />
        </div>
        <div
          class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
          style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
        >
          <div
            class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
            data-testid="card-text"
          >
            4
            <img
              alt="4 of Clubs"
              class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
              src="Clubs.svg"
            />
          </div>
          <img
            alt="4 of Clubs"
            class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
            data-testid="card-symbol"
            src="Clubs.svg"
          />
        </div>
        <div
          class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
          style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
        >
          <div
            class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
            data-testid="card-text"
          >
            5
            <img
              alt="5 of Clubs"
              class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
              src="Clubs.svg"
            />
          </div>
          <img
            alt="5 of Clubs"
            class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
            data-testid="card-symbol"
            src="Clubs.svg"
          />
        </div>
        <div
          class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
          style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
        >
          <div
            class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
            data-testid="card-text"
          >
            6
            <img
              alt="6 of Clubs"
              class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
              src="Clubs.svg"
            />
          </div>
          <img
            alt="6 of Clubs"
            class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
            data-testid="card-symbol"
            src="Clubs.svg"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex my-6 md:px-8 md:my-32 flex-wrap gap-x-3 gap-y-5 md:gap-5 justify-center w-full"
    >
      <div
        class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
        style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
      >
        <div
          class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
          data-testid="card-text"
        >
          2
          <img
            alt="2 of Clubs"
            class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
            src="Clubs.svg"
          />
        </div>
        <img
          alt="2 of Clubs"
          class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
          data-testid="card-symbol"
          src="Clubs.svg"
        />
      </div>
      <div
        class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
        style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
      >
        <div
          class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
          data-testid="card-text"
        >
          3
          <img
            alt="3 of Clubs"
            class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
            src="Clubs.svg"
          />
        </div>
        <img
          alt="3 of Clubs"
          class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
          data-testid="card-symbol"
          src="Clubs.svg"
        />
      </div>
      <div
        class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
        style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
      >
        <div
          class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
          data-testid="card-text"
        >
          4
          <img
            alt="4 of Clubs"
            class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
            src="Clubs.svg"
          />
        </div>
        <img
          alt="4 of Clubs"
          class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
          data-testid="card-symbol"
          src="Clubs.svg"
        />
      </div>
      <div
        class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
        style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
      >
        <div
          class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
          data-testid="card-text"
        >
          5
          <img
            alt="5 of Clubs"
            class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
            src="Clubs.svg"
          />
        </div>
        <img
          alt="5 of Clubs"
          class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
          data-testid="card-symbol"
          src="Clubs.svg"
        />
      </div>
      <div
        class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
        style="transition: all 300ms ease-in-out; transform-origin: 50% 0; opacity: 1; transform: translateY(20px);"
      >
        <div
          class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
          data-testid="card-text"
        >
          6
          <img
            alt="6 of Clubs"
            class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
            src="Clubs.svg"
          />
        </div>
        <img
          alt="6 of Clubs"
          class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
          data-testid="card-symbol"
          src="Clubs.svg"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
